zebrad/src/commands/seed.rs

```diff
@@ -0,0 +1,197 @@
+//! `seed` subcommand - runs a dns seeder
+
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use abscissa_core::{config, Command, FrameworkError, Options, Runnable};
+use futures::channel::oneshot;
+use tower::{buffer::Buffer, Service, ServiceExt};
+use zebra_network::{AddressBook, BoxedStdError, Request, Response};
+
+use crate::{config::ZebradConfig, prelude::*};
+
+/// Whether our `SeedService` is poll_ready or not.
+#[derive(Debug)]
+enum SeederState {
+    // This is kinda gross but ¯\_(ツ)_/¯
+    TempState,
+    /// Waiting for the address book to be shared with us via the oneshot channel.
+    AwaitingAddressBook(oneshot::Receiver<Arc<Mutex<AddressBook>>>),
+    /// Address book received, ready to service requests.
+    Ready(Arc<Mutex<AddressBook>>),
+}
+
+#[derive(Debug)]
+struct SeedService {
+    state: SeederState,
+}
+
+impl Service<Request> for SeedService {
+    type Response = Response;
+    type Error = BoxedStdError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        debug!("SeedService.state: {:?}", self.state);
+
+        let mut poll_result = Poll::Pending;
+
+        // We want to be able to consume the state, but it's behind a mutable
+        // reference, so we can't move it out of self without swapping in a
+        // placeholder, even if we immediately overwrite the placeholder.
+        let tmp_state = std::mem::replace(&mut self.state, SeederState::TempState);
```

Comment:
The `TempState` variant just exists to deal with this awkward ownership thing, right?

Reply:
Correct

Reply:
I'm not immediately sure of a better thing to do but I think it would be good to see if it's possible to avoid it...